segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/stat/SegmentStatsRecorderImpl.java

```diff
@@ -186,23 +176,17 @@ public void record(String streamSegmentName, long dataLength, int numOfEvents) {
     @Override
     public void merge(String streamSegmentName, long dataLength, int numOfEvents, long txnCreationTime) {
         SegmentAggregates aggregates = getSegmentAggregate(streamSegmentName);
-        if (aggregates != null) {
-            aggregates.updateTx(dataLength, numOfEvents, txnCreationTime);
+        if (aggregates != null && aggregates.updateTx(dataLength, numOfEvents, txnCreationTime)) {
             report(streamSegmentName, aggregates);
         }
     }
 
-    private long report(String streamSegmentName, SegmentAggregates aggregates) {
-        return aggregates.lastReportedTime.getAndUpdate(prev -> {
-            if (System.currentTimeMillis() - prev > reportingDuration) {
-                reporter.report(streamSegmentName,
-                        aggregates.getTargetRate(), aggregates.getScaleType(), aggregates.getStartTime(),
-                        aggregates.getTwoMinuteRate(), aggregates.getFiveMinuteRate(),
-                        aggregates.getTenMinuteRate(), aggregates.getTwentyMinuteRate());
-                return System.currentTimeMillis();
-            }
-
-            return prev;
+    private void report(String streamSegmentName, SegmentAggregates aggregates) {
+        aggregates.reportIfNecessary(reportingDuration, () -> {
+            reporter.report(streamSegmentName,
```

Comment:
My understanding is that this has little to do with the rest of the data in SegmentAggregates. It just needs some place to store the `lastReportTime` for this segment so that it would execute conditionally if too much time elapsed since then. I tried to find a new home for this pattern but I would have ended up needing yet another Map/Cache for Segments to Timestamps, so I preferred to just leave it as it was. 

The lock taken while executing this is different than the lock used to synchronize the `get*Rate` methods, so they will not interfere.

Reply:
Yes, I agree. I have a subsequent PR built on top of this (I'll file when we merge this one) that tackles this problem as well. #3332